### ISSUE TITLE
Review impact of global using directives on existing rules

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#3982]
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-directive

**Comment by sharwell** [comment]
Audit by Deep Research

# Impacted Ordering Rules (Using Directives) 

C# 10 introduces **global using** directives. These must be handled carefully by ordering and layout rules so the analyzers:

* Ignore valid global usings,
* Enforce correct ordering **within** global usings,
* And still enforce rules for normal (non-global) usings.

This document lists the impacted rules (and one layout rule) and summarizes the behavior.

---

## SA1200 – UsingDirectivesMustBePlacedCorrectly

### Behavior with global usings

* Global using directives are **always** top-level by language design (outside any namespace).
* Historically, SA1200 reported a violation for any using outside a namespace.
* This caused false positives for global usings.

### Required behavior

* SA1200 **must not** report diagnostics for `global using` directives.
* Implementation skips global usings when collecting “outside-namespace” usings.
* Documentation clarifies that **global using directives are exempt** from SA1200.

### Tests

* Files with global usings before a namespace:

  ```csharp
  global using System;
  global using MyLib;

  namespace Example;
  ```

  No SA1200 diagnostics.

* Files using top-level statements with global usings.

---

## SA1208 – SystemUsingDirectivesMustBePlacedBeforeOtherUsingDirectives

### Behavior with global usings

* SA1208 requires `System.*` usings to appear before non-`System` usings.

* With global usings, we effectively have two independent groups:

  * Global usings.
  * Non-global (normal) usings.

* The rule should enforce **System-first** order **within each group**, but **not across** global/non-global boundaries.

### Required behavior

* Only compare usings that share the same `global`-ness:

  * Global vs global.
  * Non-global vs non-global.
* Example:

  * **Violation** (global group):

    ```csharp
    global using MyLib;
    global using System;
    ```

  * **No violation** (global vs local):

    ```csharp
    global using System;

    using MyLib;
    ```

  Even if the local `using MyLib;` appears before local `using System;`, it is still checked separately for SA1208 in the **local** group.

### Tests

* Mixed files with global and local usings:

  * Check that each group enforces `System.*` before other namespaces.
  * Confirm no cross-group comparisons.

---

## SA1209 – UsingAliasDirectivesMustBePlacedAfterOtherUsingDirectives

### Behavior with global usings

* Aliased usings (`using X = Y;`) must appear after non-alias usings.
* With global usings:

  * Global alias usings must come after global non-alias usings.
  * Local alias usings must come after local non-alias usings.
* No comparisons should be made **between** global and non-global directives.

### Required behavior

* Only compare alias ordering within the same global/local context.
* Example:

  * **Allowed**:

    ```csharp
    global using System;
    global using Alias = System.Console;

    using System;
    using Alias2 = System.Console;
    ```

  * **Violation examples**:

    * Global alias before global regular.
    * Local alias before local regular.

### Tests

* Global-only scenarios.
* Local-only scenarios.
* Combined global + local scenarios.

---

## SA1210 – UsingDirectivesMustBeOrderedAlphabeticallyByNamespace

### Behavior with global usings

* SA1210 enforces alphabetical ordering of usings by namespace.
* With C# 10, the ordering must be evaluated:

  * **Separately for global usings**.
  * **Separately for local usings**.

### Required behavior

* Global usings form their own ordered block.
* Local usings form another ordered block.
* The analyzer should not compare global and local usings to each other for alphabetical order.

### Tests

* Files with a mix of global and local usings:

  * Confirm no SA1210 diagnostics if each block is internally sorted.
  * Confirm misordered entries **within** each block are still reported.

---

## SA1211 – UsingAliasDirectivesMustBeOrderedAlphabeticallyByAliasName

### Behavior with global usings

* SA1211 orders `using Alias = ...` directives by alias name.
* Similar to SA1210:

  * Alphabetical checking applies separately to:

    * Global alias usings.
    * Local alias usings.

### Required behavior

* Only compare alias names within the same global/local group.
* Ignore ordering relationship between a global alias and a local alias.

### Tests

* Separate sets of global and local alias usings, each requiring alphabetic sorting.
* Ensure out-of-order alias in global block is flagged, but out-of-order relative to local block is ignored.

---

## SA1216 – UsingStaticDirectivesMustBePlacedAtTheCorrectLocation

### Behavior with global usings

* SA1216 enforces this within a group:

  * Normal usings.
  * Then `using static`.
  * Then alias usings.

* With global usings, we essentially have two sequences:

  * Global: regular → static → alias.
  * Local: regular → static → alias.

### Required behavior

* The rule must reset its internal state when moving between global and non-global sections.
* Only order **within** each set:

  * Global static usings must appear between global regular usings and global alias usings.
  * Local static usings must appear between local regular usings and local alias usings.

### Tests

* **Violation**:

  ```csharp
  global using static System.Console;
  global using System;
  ```

* **No violation** when a global static using appears before **local** regular usings (different group).

---

## SA1217 – UsingStaticDirectivesMustBeOrderedAlphabetically

### Behavior with global usings

* SA1217 requires `using static ...` directives to be alphabetized by fully-qualified type name.
* With global usings, global static usings and local static usings must be handled separately.

### Required behavior

* Alphabetical checking occurs independently for:

  * Global static usings.
  * Local static usings.
* No cross-group comparisons.

### Tests

* A block of global static usings, intentionally misordered, should trigger SA1217.
* Mixed global/local static usings out of order **relative to each other** must not cause SA1217 diagnostics.

---

## Impacted Layout Rules

### SA1516 – ElementsMustBeSeparatedByBlankLine

SA1516 controls blank lines between elements and can also enforce spacing between groups of usings (`blankLinesBetweenUsingGroups` in `stylecop.json`).

Global usings introduce a new top-level group:

* The analyzer must treat transitions between:

  * Global usings,
  * Namespace-level usings,
  * Other code,
    as potential group boundaries for blank-line rules.

### Required behavior

* Using directives are already grouped by type (System, regular, static, alias). This grouping now implicitly respects global vs non-global lists.
* When `blankLinesBetweenUsingGroups` is enabled, the analyzer should:

  * Consider the end of the global using block as a distinct boundary.
  * Apply normal spacing rules between:

    * Global usings and file-scoped namespaces.
    * Global usings and top-level statements.
    * Global usings and namespace-level usings.

### Tests

* Files that contain:

  * Only global usings.
  * Global usings followed by file-scoped namespace or top-level statements.
  * Global usings followed by namespace-level usings.

* Ensure:

  * Required blank lines are enforced.
  * No duplicate or spurious diagnostics occur (including cases affected by earlier compiler quirks around global using lists).

---

## Other Rule Categories

* Naming, documentation, maintainability, and most readability/spacing rules are **unaffected** by global using directives.
* Example: **SA1135** (using directives must be qualified) already treats any using outside a namespace as “qualified” and thus does not require changes for global usings.

All necessary adjustments are confined to the ordering and layout rules listed above.

### Sources

* SA1200–SA1217 documentation and analyzers.
* `SA1200UsingDirectivesMustBePlacedCorrectly.cs`
* `SA1208SystemUsingDirectivesMustBePlacedBeforeOtherUsingDirectives.cs`
* `SA1216UsingStaticDirectivesMustBePlacedAtTheCorrectLocation.cs`
* `SA1516CSharp11UnitTests.cs`
* `Configuration.md` and individual `SA1200`–`SA1217` docs.
